### PR TITLE
Hover for attribute values using XSD

### LINKED ISSUE #12: [Request] Document Hover
Support hover over 
- [x] Tag
- [x] Attributes
- [ ] Values

### LINKED ISSUE #339: Self closing tag causes NPE in cvc_complex_type_2_1CodeAction.doCodeAction
At the root of lsp4xml (with [money.xsd](https://github.com/angelozerr/lsp4xml/pull/337/files#diff-3e61076f7173cbb44d7d2240e14ddcd3) present), create a file with :
```xml
<money xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="org.eclipse.lsp4xml/src/test/resources/xsd/money.xsd" currency="euros"> </money>
```
When self closing the money tag, an NPE is thrown:

```
Error - 3:22:55 p.m.] Apr 03, 2019 03:22:55 org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError()
Message: Internal error: java.lang.NullPointerException
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.NullPointerException
	at org.eclipse.lsp4xml.extensions.contentmodel.participants.codeactions.cvc_complex_type_2_1CodeAction.doCodeAction(cvc_complex_type_2_1CodeAction.java:42)
	at org.eclipse.lsp4xml.extensions.contentmodel.participants.ContentModelCodeActionParticipant.doCodeAction(ContentModelCodeActionParticipant.java:45)
	at org.eclipse.lsp4xml.services.XMLCodeActions.doCodeActions(XMLCodeActions.java:43)
	at org.eclipse.lsp4xml.services.XMLLanguageService.doCodeActions(XMLLanguageService.java:188)
	at org.eclipse.lsp4xml.XMLTextDocumentService.lambda$codeAction$15(XMLTextDocumentService.java:340)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	... 6 more

```

Apr-03-2019 15-23-57

### PR DESCRIPTION
**Pull request by NikolasKomonen**
Fixes #12
Fixes #339

### CODE REVIEW

**Comment by fbricon** [comment]
Please fix the commit message s/attribuite/attribute

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java [hunk 1]
```diff
@@ -90,6 +105,67 @@ private XSObjectList getAnnotations() {
...
+	private XSObjectList getValueAnnotations() {
+		// Try get xs:annotation from the element declaration
+		XSAttributeDeclaration attributeDeclaration = getAttrDeclaration();
+		XSSimpleTypeDefinition simpleTypeDefinition = attributeDeclaration.getTypeDefinition();
+		XSSimpleTypeDecl simpleTypeDecl;
+		
+	
+		XSObjectList annotation = null; // The XSD tag that holds the documentation tag
+
+		if(simpleTypeDefinition != null && simpleTypeDefinition instanceof XSSimpleTypeDecl) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java]: simpleTypeDefinition != null  is redundant

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java [hunk 2]
```diff
@@ -90,6 +105,67 @@ private XSObjectList getAnnotations() {
...
+		if (typeDefinition.getTypeCategory() == XSTypeDefinition.COMPLEX_TYPE) {
+			return ((XSComplexTypeDecl) typeDefinition).getAnnotations();
+		} else if (typeDefinition.getTypeCategory() == XSTypeDefinition.SIMPLE_TYPE) {
+			return ((XSSimpleTypeDecl) typeDefinition).getAnnotations();
+		}
+		return null;
+	}
+
+	private boolean onlyContainsNull(Object[] arr) {
+		if(arr == null) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java]: arr == null || arr.length == 0

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java [hunk 3]
```diff
@@ -90,6 +105,67 @@ private XSObjectList getAnnotations() {
...
+		
+	
+		XSObjectList annotation = null; // The XSD tag that holds the documentation tag
+
+		if(simpleTypeDefinition != null && simpleTypeDefinition instanceof XSSimpleTypeDecl) {
+			simpleTypeDecl = (XSSimpleTypeDecl) simpleTypeDefinition;
+			XSObjectList tempAnnotation = simpleTypeDecl.getMultiValueFacets();
+			XSMultiValueFacet zz = (XSMultiValueFacet) tempAnnotation.get(0);
+			tempAnnotation = zz.getAnnotations();
+			Object[] annotationArray = tempAnnotation.toArray();
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java]: do you really get back arrays containing null-elements?

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/XSDAnnotationModel.java [hunk 4]
```diff
@@ -47,12 +51,37 @@ public String getDocumentation() {
...
 		for (Object object : annotations) {
-			XSAnnotation annotation = (XSAnnotation) object;
+			XSAnnotation annotation = null;
+			if(object instanceof XSMultiValueFacet && value != null) {
+				XSMultiValueFacet multiValueFacet = (XSMultiValueFacet) object;
+				ObjectList enumerationValues = multiValueFacet.getEnumerationValues();
+				XSObjectList annotationValues = multiValueFacet.getAnnotations();
+				for (int i = 0; i < enumerationValues.getLength(); i++) {
+					Object enumValue = enumerationValues.get(i);
+					String enumString = null;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/XSDAnnotationModel.java]: String enumString  = ((ValidatedInfo) enumValue).stringValue();

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java (hunk 3 above)
Comment (NikolasKomonen) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDAttributeDeclaration.java]: yes

**Comment by NikolasKomonen** [comment]
@fbricon updated

**Comment by fbricon** [comment]
If you open money.xsd and hover over any value of the attributes there, an exception is thrown:

```
Message: While performing IHoverParticipant#onTag
java.lang.IndexOutOfBoundsException: Index: 0
	at org.apache.xerces.impl.xs.util.XSObjectListImpl.get(Unknown Source)
	at org.eclipse.lsp4xml.extensions.xsd.contentmodel.CMXSDAttributeDeclaration.getValueAnnotations(CMXSDAttributeDeclaration.java:131)
	at org.eclipse.lsp4xml.extensions.xsd.contentmodel.CMXSDAttributeDeclaration.getValueDocumentation(CMXSDAttributeDeclaration.java:75)
	at org.eclipse.lsp4xml.extensions.contentmodel.participants.ContentModelHoverParticipant.onAttributeValue(ContentModelHoverParticipant.java:101)
	at org.eclipse.lsp4xml.services.XMLHover.getAttrValueHover(XMLHover.java:162)
	at org.eclipse.lsp4xml.services.XMLHover.doHover(XMLHover.java:75)
	at org.eclipse.lsp4xml.services.XMLLanguageService.doHover(XMLLanguageService.java:109)
	at org.eclipse.lsp4xml.XMLTextDocumentService.lambda$hover$2(XMLTextDocumentService.java:197)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Pull request merged**